File: python/hackerrank/algorithms/implementation/find-digits.py
'''
https://www.hackerrank.com/challenges/find-digits

Number of digit of N that evenly divide N

Sample Input
2
12
1012

Sample Output
2
3

'''

def find_num_digits_dividing(N):
	num_of_digits = 0
	rest_of_n = N

	while rest_of_n != 0:
		digit = rest_of_n % 10
		rest_of_n = rest_of_n / 10

		try:
			if (N % digit == 0):
				num_of_digits += 1
		except ZeroDivisionError:
			continue


	return num_of_digits



n = int(raw_input())
for i in xrange(n):
	N = int(raw_input())
	print find_num_digits_dividing(N)

File: haskell/hackerrank/algorithms/implementation/find-digits.hs
{-
https://www.hackerrank.com/challenges/find-digits

Number of digit of N that evenly divide N

Sample Input
2
12
1012

Sample Output
2
3
-}

find_digits_dividing' :: Int -> Int -> Int -> Int
find_digits_dividing' _ 0 sum = sum
find_digits_dividing' num rest_of_num sum = 
	let digit = rest_of_num `mod` 10
	in
	if (digit == 0) || (num `mod` digit /= 0)
			then
				find_digits_dividing' num (rest_of_num `div` 10) sum
			else 
				find_digits_dividing' num (rest_of_num `div` 10) (sum + 1)

find_digits_dividing :: Int -> Int
find_digits_dividing num = 
	find_digits_dividing' num num 0



-- Read n lines and process separately
getLinesAndProcess :: Int -> IO ()
getLinesAndProcess 0 = return ()
getLinesAndProcess n = do
	num <- readLn :: IO Int
	putStrLn ( show (find_digits_dividing num) )
	getLinesAndProcess (n-1)

main = do
	n <- readLn :: IO Int
	getLinesAndProcess n

